### DISCUSSION TITLE
SQLAlchemy/PostgreSQL: Generate a Unique/Short Primary Key

### DISCUSSION THREAD

**Discussion by thisiseddy-ab** [sqlalchemy/sqlalchemy#9244] [Usage Questions] [answered]
I was inspired from [this][1] library to create a short primary key like youtubes 
  
Than i found [this][2] library that converts uuid in short,base64,url-safe id,my question is what approach should i take to save it in database

#### First Approach (Saving like a string primare key)
```py
from uuid import uuid4
from b64uuid import B64UUID

def create_id():
    return str(B64UUID(uuid4()))

class User(Base):
    __tablename__ = "user"
    id = Column('id', String(200),primary_key=True, default=create_id)
    firstname = Column('firstname', String(200))
    lastname = Column('lastname', String(200))
````

#### Second Approach (Saving like UUID and then in REST API convert it  back and forth the uuid to base64)
```py
from uuid import uuid4
from sqlalchemy.dialects.postgresql import UUID

class User_UUID(Base):
    __tablename__ = "user_uuid"
    id = Column(UUID(as_uuid=True),primary_key=True,index=True,default=uuid4)
    firstname = Column('firstname', String(200))
    lastname = Column('lastname', String(200))
````

```py
@app.get("/uuid_get_users")
def suuid_get_users():
    with session_scope() as session:
        users = session.query(User_UUID).all()
        if users != None:
            return_body = []
            for user in users:
                return_body.append(
                    {
                        "id" : str(B64UUID(user.id)),
                        "firstname" : user.firstname,
                        "lastname" : user.lastname,
                    }
                )
            return return_body


@app.get("/uuid_get_user/{id}")
def uuid_get_userr(id):
    with session_scope() as session:
        uuid = B64UUID(id).uuid
        user = session.query(User_UUID).get(uuid)
        if user != None:    
            return_body = {
                "id" : str(B64UUID(user.id)),
                "firstname" : user.firstname,
                "lastname" : user.lastname,
            }
            return return_body
````

My question is which approach is more efficient,i don't know are string primarykeys efficient,Hussein Nasser in [this][3] video says uuids are efficient in PostgreSQL because really are like secondary keys.I wouldn't suggest this approach in MySQL,is someone with knowledge in PostgreSQL to help me.

  [1]: https://github.com/turbo/pg-shortkey
  [2]: https://pypi.org/project/b64uuid/
  [3]: https://www.youtube.com/watch?v=Y5mWz4vK10A

**Comment by zzzeek** [comment, marked as answer]
PostgreSQL's UUID type is a native type to PostgreSQL, it is definitely more optimized than using a plain string, so go with that.

**Reply by thisiseddy-ab** [reply]
That's what i was thinking,bu hear me out what if chose to change the REST API language and build it with Javascript i don't know that all bas64 library's produce the same id.
In that case u can make a service in python , that just converts base64 id to UUID.
Your suggestion sounds better,thanks.
